test/e2e/storage/vsphere/vsphere_volume_vsan_policy.go

```diff
@@ -211,6 +207,8 @@ var _ = utils.SIGDescribe("Storage Policy Based Volume Provisioning [Feature:vsp
 		scParameters[Datastore] = vsanDatastore
 		framework.Logf("Invoking test for SPBM storage policy: %+v", scParameters)
 		kubernetesClusterName := GetAndExpectStringEnvVar(KubernetesClusterName)
+		masterNode, err := getControlPlaneNode(client)
+		framework.ExpectNoError(err)
 		invokeStaleDummyVMTestWithStoragePolicy(client, masterNode, namespace, kubernetesClusterName, scParameters)
```

Comment:
i don't really understand why this function really needs a "master" node, though.
it just ends up calling:

```
nodeInfo := TestContext.NodeMapper.GetNodeInfo(masterNode)
```

Reply:
@divyenpatel

you added the "master" node related logic here:
https://github.com/kubernetes/kubernetes/pull/59761
the PR and tracking issue did not explain directly why we need to detect if a node is "master" or not.

any comments before we merge the change in this PR?

Reply:
@oomichi 

> TBH I am still not sure what is this test purpose.
> Especially what is invokeStaleDummyVMTestWithStoragePolicy() doing.

With in-tree VCP, when the volume is created using storage policy, shadow (dummy) VM is getting created and deleted to apply SPBM policy on the volume.

https://github.com/kubernetes/legacy-cloud-providers/blob/master/vsphere/vclib/diskmanagers/vmdm.go#L157-L161

But when for some reason if VM deletion fails, we just print an error message and continue with provisioning workflow.

We have a routine in VCP which runs every 5 minutes and helps clean up these leftover VMs. 
https://github.com/kubernetes/legacy-cloud-providers/blob/master/vsphere/vsphere_util.go#L382-L414

The test is helping to validate the functionality of this routine.

Reply:
@divyenpatel does it really need a "master" node?

https://github.com/kubernetes/kubernetes/blob/5e5defb9b5df92fccc2a2a8efbb956c6c58ae38f/test/e2e/storage/vsphere/vsphere_volume_vsan_policy.go#L337-L338

is `DataCenterRef` only set for "master" nodes?

Reply:
@neolit123 

Basically we need a datacenter specified in the Workspace, thats where dummy VM is getting created.

```
[Global]
user = "administrator@vsphere.local"
password = "password"
port = "443"
insecure-flag = "1"

[VirtualCenter "10.0.1.200"]
datacenters = "DC-1"

[Workspace]
server = "10.0.1.200"
datacenter = "DC-1"
default-datastore = "vsanDatastore"
resourcepool-path = "ClusterNameHere/Resources"
folder = "kubernetes"

[Disk]
scsicontrollertype = pvscsi
```

[Workspace] - Defines the temporary location of shadow VMs provisioned to create volumes

    server is the vCenter Server IP or FQDN
    datacenter - the Datacenter to provision temporary VMs to for volume provisioning

Refer - https://vmware.github.io/vsphere-storage-for-kubernetes/documentation/existing.html

Reply:
are you happy with the change in this PR to change the logic of detecting "master" (control-plane) nodes?

Reply:
We do not need to get master node for `invokeStaleDummyVMTestWithStoragePolicy` test. We need to read `vsphere.conf` file supplied in the env variable and get workspace datacenter.

Internally in the test, this is how it is used - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/vsphere/vsphere_volume_vsan_policy.go#L339-L340

This is wrong. 

The test will be able to validate functionality in the case when the master node is present on the datacenter configured in the workspace.

but since this test issue was already present, I will not block this PR.

@oomichi Can you put a note in the doc for this test- "invokeStaleDummyVMTestWithStoragePolicy assumes master node is present on the datacenter specified in the workspace section of vsphere.conf file"

Reply:
@divyenpatel 
> With in-tree VCP, when the volume is created using storage policy, shadow (dummy) VM is getting created and deleted to apply SPBM policy on the volume.

Thanks for your explanation. I got the point.

> @oomichi Can you put a note in the doc for this test- "invokeStaleDummyVMTestWithStoragePolicy assumes master node is present on the datacenter specified in the workspace section of vsphere.conf file"

Is it fine to put the above note on this PR message?
Or should we put it on `invokeStaleDummyVMTestWithStoragePolicy()` itself?

Reply:
would be good to put this as func doc for `invokeStaleDummyVMTestWithStoragePolicy`.

Reply:
@divyenpatel Thanks, I got it.